index.js

```diff
@@ -1,155 +1,154 @@
 'use strict';
 
-module.exports = Plugin;
-function Plugin(inputNodes, options) {
-  if (!(this instanceof Plugin)) throw new TypeError('Missing `new` operator');
-  // Remember current call stack (minus "Error" line)
-  this._instantiationStack = new Error().stack.replace(/[^\n]*\n/, '');
-
-  options = options || {};
-  if (options.name != null) {
-    this._name = options.name;
-  } else if (this.constructor && this.constructor.name != null) {
-    this._name = this.constructor.name;
-  } else {
-    this._name = 'Plugin';
-  }
-  this._annotation = options.annotation;
-
-  let label = this._name + (this._annotation != null ? ' (' + this._annotation + ')' : '');
-  if (!Array.isArray(inputNodes))
-    throw new TypeError(
-      label + ': Expected an array of input nodes (input trees), got ' + inputNodes
-    );
-  for (let i = 0; i < inputNodes.length; i++) {
-    if (!isPossibleNode(inputNodes[i])) {
+module.exports = class Plugin {
+  constructor(inputNodes, options) {
+    // Remember current call stack (minus "Error" line)
+    this._instantiationStack = new Error().stack.replace(/[^\n]*\n/, '');
+
+    options = options || {};
+    if (options.name != null) {
+      this._name = options.name;
+    } else if (this.constructor && this.constructor.name != null) {
+      this._name = this.constructor.name;
+    } else {
+      this._name = 'Plugin';
+    }
+    this._annotation = options.annotation;
+
+    let label = this._name + (this._annotation != null ? ' (' + this._annotation + ')' : '');
+    if (!Array.isArray(inputNodes))
       throw new TypeError(
-        label + ': Expected Broccoli node, got ' + inputNodes[i] + ' for inputNodes[' + i + ']'
+        label + ': Expected an array of input nodes (input trees), got ' + inputNodes
       );
+    for (let i = 0; i < inputNodes.length; i++) {
+      if (!isPossibleNode(inputNodes[i])) {
+        throw new TypeError(
+          label + ': Expected Broccoli node, got ' + inputNodes[i] + ' for inputNodes[' + i + ']'
+        );
+      }
     }
-  }
 
-  this._baseConstructorCalled = true;
-  this._inputNodes = inputNodes;
-  this._persistentOutput = !!options.persistentOutput;
-  this._needsCache = options.needsCache != null ? !!options.needsCache : true;
+    this._baseConstructorCalled = true;
+    this._inputNodes = inputNodes;
+    this._persistentOutput = !!options.persistentOutput;
+    this._needsCache = options.needsCache != null ? !!options.needsCache : true;
 
-  this._checkOverrides();
-}
+    this._checkOverrides();
 
-Plugin.prototype._checkOverrides = function() {
-  if (typeof this.rebuild === 'function') {
-    throw new Error('For compatibility, plugins must not define a plugin.rebuild() function');
-  }
-  if (this.read !== Plugin.prototype.read) {
-    throw new Error('For compatibility, plugins must not define a plugin.read() function');
+    // For future extensibility, we version the API using feature flags
+    this.__broccoliFeatures__ = Object.freeze({
```

Comment:
We can actually create a `const FEATURES` singleton at the top of the module, and share it between all instances. No need to create a new object for all of these.

Reply:
Good catch. Updated